internal/worker/concurrency_auto_scaler.go

```diff
@@ -0,0 +1,263 @@
+	*/
+	var shouldEnable bool
+	if hint.EnableAutoConfig != nil && *hint.EnableAutoConfig {
+		shouldEnable = true
+	}
+	if switched := c.enable.CompareAndSwap(!shouldEnable, shouldEnable); switched {
+		if shouldEnable {
+			c.logEvent(autoScalerEventEnable)
+		} else {
+			c.resetConcurrency()
+			c.logEvent(autoScalerEventDisable)
+		}
+	}
+}
+
+// resetConcurrency reset poller quota to the max value. This will be used for gracefully switching the auto scaler off to avoid workers stuck in the wrong state
+func (c *ConcurrencyAutoScaler) resetConcurrency() {
+	c.concurrency.PollerPermit.SetQuota(c.pollerInitCount)
+}
+
+func (c *ConcurrencyAutoScaler) logEvent(event autoScalerEvent) {
+	if c.enable.Load() {
+		c.scope.Counter("concurrency_auto_scaler.enabled").Inc(1)
+	} else {
+		c.scope.Counter("concurrency_auto_scaler.disabled").Inc(1)
+	}
+	c.scope.Gauge("poller_in_action").Update(float64(c.concurrency.PollerPermit.Count()))
+	c.scope.Gauge("poller_quota").Update(float64(c.concurrency.PollerPermit.Quota()))
+	c.scope.Gauge("poller_wait_time").Update(math.Exp2(c.pollerWaitTimeInMsLog2.Average()))
+	c.log.Debug(autoScalerEventLogMsg,
+		zap.Time("time", c.clock.Now()),
+		zap.String("event", string(event)),
+		zap.Bool("enabled", c.enable.Load()),
+		zap.Int("poller_quota", c.concurrency.PollerPermit.Quota()),
+		zap.Int("poller_in_action", c.concurrency.PollerPermit.Count()),
+	)
+}
+
+func (c *ConcurrencyAutoScaler) updatePollerPermit() {
+	if !c.enable.Load() { // skip update if auto scaler is disabled
+		c.logEvent(autoScalerEventPollerSkipUpdateNotEnabled)
+		return
+	}
+	updateTime := c.clock.Now()
+	if updateTime.Before(c.pollerPermitLastUpdate.Add(c.cooldown)) { // before cooldown
+		c.logEvent(autoScalerEventPollerSkipUpdateCooldown)
+		return
+	}
+	currentQuota := c.concurrency.PollerPermit.Quota()
+	newQuota := int(math.Round(float64(currentQuota) * c.pollerWaitTimeInMsLog2.Average() / targetPollerWaitTimeInMsLog2))
+	if newQuota < c.pollerMinCount {
+		newQuota = c.pollerMinCount
+	}
+	if newQuota > c.pollerMaxCount {
+		newQuota = c.pollerMaxCount
+	}
+	if newQuota == currentQuota {
+		c.logEvent(autoScalerEventPollerSkipUpdateNoChange)
+		return
+	}
+	c.concurrency.PollerPermit.SetQuota(newQuota)
+	c.pollerPermitLastUpdate = updateTime
+	c.logEvent(autoScalerEventPollerUpdate)
+}
+
+type rollingAverage struct {
+	mu     sync.RWMutex
+	window []float64
+	index  int
+	sum    float64
+	count  int
+}
+
+func newRollingAverage(capacity int) *rollingAverage {
+	return &rollingAverage{
+		window: make([]float64, capacity),
+	}
+}
+
+// Add always add positive numbers
+func (r *rollingAverage) Add(value float64) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	// no op on zero rolling window
+	if len(r.window) == 0 {
+		return
+	}
+
+	// replace the old value with the new value
+	r.index %= len(r.window)
+	r.sum += value - r.window[r.index]
+	r.window[r.index] = value
+	r.index++
```

Comment:
general design critique: where possible, make sure all state valid when it leaves the function boundaries.

e.g. in this case, r.index is allowed to be ==len (and unsafe to use) every `len(r.window)` times it's called.  if this were instead:
```suggestion
	r.sum += value - r.window[r.index]
	r.window[r.index] = value
	r.index++
	r.index %= len(r.window)
```
it'd always be correct.